test/core/client_channel/lb_policy/round_robin_test.cc

```diff
@@ -0,0 +1,373 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#include <stddef.h>
+
+#include <algorithm>
+#include <map>
+#include <memory>
+#include <utility>
+#include <vector>
+
+#include "absl/status/status.h"
+#include "absl/status/statusor.h"
+#include "absl/strings/string_view.h"
+#include "gtest/gtest.h"
+
+#include <grpc/grpc.h>
+
+#include "src/core/ext/filters/client_channel/subchannel_pool_interface.h"
+#include "src/core/lib/channel/channel_args.h"
+#include "src/core/lib/gprpp/orphanable.h"
+#include "src/core/lib/iomgr/resolved_address.h"
+#include "src/core/lib/load_balancing/lb_policy.h"
+#include "src/core/lib/resolver/server_address.h"
+#include "test/core/client_channel/lb_policy/lb_policy_test_lib.h"
+#include "test/core/util/test_config.h"
+
+namespace grpc_core {
+namespace testing {
+namespace {
+
+class RoundRobinTest : public LoadBalancingPolicyTest {
+ public:
+  LoadBalancingPolicy::UpdateArgs BuildUpdateArgs(
+      absl::Span<const std::string> subchannel_addresses) {
+    LoadBalancingPolicy::UpdateArgs update_args;
+    update_args.addresses.emplace();
+    for (const auto& addr : subchannel_addresses) {
+      update_args.addresses->emplace_back(MakeAddress(addr), ChannelArgs());
+    }
+    return update_args;
+  }
+
+  void ExpectAllChannelsConnecting() {
+    for (size_t i = 0; i < subchannel_pool_.size(); ++i) {
+      ExpectState(GRPC_CHANNEL_CONNECTING);
+    }
+  }
+
+  // Picker should return each address in any order.
+  void ExpectPickAddresses(LoadBalancingPolicy::SubchannelPicker* picker,
+                           absl::Span<const std::string> uris,
+                           size_t iterations_per_uri = 3,
+                           SourceLocation location = SourceLocation()) {
+    std::unordered_map<std::string, int> reported_uris;
+    for (size_t i = 0; i < iterations_per_uri * uris.size(); ++i) {
+      auto address = ExpectPickAddress(picker);
+      ASSERT_TRUE(address.has_value())
+          << location.file() << ":" << location.line();
+      reported_uris[*address]++;
+    }
+    EXPECT_EQ(reported_uris.size(), uris.size())
+        << location.file() << ":" << location.line();
+    for (const std::string& uri : uris) {
+      EXPECT_EQ(reported_uris[uri], iterations_per_uri)
+          << "Subchannel " << uri << location.file() << ":" << location.line();
+    }
+  }
+};
+
+TEST_F(RoundRobinTest, SingleChannel) {
+  std::string uri = "ipv4:127.0.0.1:441";
+  auto policy = MakeLbPolicy("round_robin");
+  auto status = ApplyUpdate(BuildUpdateArgs({uri}), policy.get());
+
```

Comment:
Nit: Please don't leave blank lines within a function.  I generally find that anywhere I really want to break up the flow, I'm better off using a one-line comment than just having a blank line.

Same thing throughout.

Reply:
Done